File: backend/app/schemas/market_mood.py
"""Provider-neutral Market Mood read contracts."""

from datetime import datetime
from typing import Literal, TypeAlias

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.services.privacy import FORBIDDEN_TRADE_REVIEW_WORKSPACE_KEYS, find_forbidden_keys


MarketMoodDataMode: TypeAlias = Literal["synthetic", "provider_reference", "unavailable"]
MarketMoodFreshnessStatus: TypeAlias = Literal["fresh", "stale", "unavailable"]
MarketMoodRating: TypeAlias = Literal["extreme_fear", "fear", "neutral", "greed", "extreme_greed", "unknown"]
MarketMoodComparisonWindow: TypeAlias = Literal["1w", "1m", "1y"]
MarketMoodRefreshStatus: TypeAlias = Literal["refreshed", "failed"]

PROHIBITED_MARKET_MOOD_PHRASES = (
    "i recommend",
    "recommendation",
    "buy ",
    "sell ",
    "risk-on",
    "risk-off",
    "safe-to-trade",
    "safe to trade",
    "ready-to-trade",
    "ready to trade",
    "guaranteed return",
    "urgent",
    "urgency",
    "place order",
    "submit order",
    "execute trade",
)


def validate_market_mood_payload(value: object) -> None:
    """Reject private fields and advice/execution wording in Market Mood responses."""

    forbidden = find_forbidden_keys(value, forbidden_keys=FORBIDDEN_TRADE_REVIEW_WORKSPACE_KEYS)
    if forbidden:
        raise ValueError(f"market mood payload contains forbidden private fields: {sorted(forbidden)}")
    rendered = repr(value).lower()
    for phrase in PROHIBITED_MARKET_MOOD_PHRASES:
        if phrase in rendered:
            raise ValueError(f"market mood payload contains prohibited wording: {phrase}")


class MarketMoodTrendPointRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    date: str
    score: float | None = None
    score_label: str | None = None
    rating: MarketMoodRating
    rating_label: str


class MarketMoodComparisonRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    window: MarketMoodComparisonWindow
    prior_score: float | None = None
    prior_score_label: str | None = None
    change_label: str | None = None
    is_available: bool


class MarketMoodComponentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    component_key: str
    display_name: str
    score: float | None = None
    score_label: str | None = None
    rating: MarketMoodRating
    rating_label: str


class MarketMoodRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    data_mode: MarketMoodDataMode
    source_label: str
    source_detail_label: str
    source_rights_notice: str
    generated_at: datetime
    updated_at_utc: datetime | None = None
    updated_at_label: str | None = None
    freshness_status: MarketMoodFreshnessStatus
    freshness_label: str
    is_trading_signal: bool
    is_actionability_input: bool
    is_risk_rule_input: bool
    score: float | None = None
    score_label: str | None = None
    score_min: int = Field(default=0)
    score_max: int = Field(default=100)
    rating: MarketMoodRating
    rating_label: str
    trend_series: tuple[MarketMoodTrendPointRead, ...]
    comparisons: tuple[MarketMoodComparisonRead, ...]
    components: tuple[MarketMoodComponentRead, ...]
    caveat_codes: tuple[str, ...]
    limitations: tuple[str, ...]
    status_message: str | None = None

    @model_validator(mode="after")
    def market_mood_read_must_be_safe(self) -> "MarketMoodRead":
        if self.is_trading_signal or self.is_actionability_input or self.is_risk_rule_input:
            raise ValueError("market mood must not be a signal, actionability input, or risk-rule input")
        if self.score_min != 0 or self.score_max != 100:
            raise ValueError("market mood score range must remain 0-100")
        validate_market_mood_payload(self.model_dump(mode="python"))
        return self


class MarketMoodRefreshStatusRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    status: MarketMoodRefreshStatus
    data_mode: MarketMoodDataMode
    source_label: str
    generated_at: datetime | None = None
    updated_at_utc: datetime | None = None
    record_count: int
    message: str

    @model_validator(mode="after")
    def refresh_status_must_be_safe(self) -> "MarketMoodRefreshStatusRead":
        validate_market_mood_payload(self.model_dump(mode="python"))
        return self

File: frontend/src/types/marketMood.ts
/**
 * TypeScript mirror of backend Market Mood read schemas (Phase 26A).
 *
 * Mirrors:
 *   backend/app/schemas/market_mood.py
 *     MarketMoodTrendPointRead, MarketMoodComparisonRead,
 *     MarketMoodComponentRead, MarketMoodIndicatorHistoryPointRead,
 *     MarketMoodIndicatorRead, MarketMoodRead, MarketMoodDetailRead,
 *     MarketMoodRefreshStatusRead
 *
 * Strictly mirror the backend. The frontend never invents fields, never parses
 * or computes financial/trading meaning, and renders backend display labels
 * verbatim. Market Mood is broad sentiment context only — not a trading signal,
 * not an actionability input, not a risk-rule input.
 */

/* ── Enum literals ─────────────────────────────────────────────────────── */

export type MarketMoodDataMode = "synthetic" | "provider_reference" | "unavailable";

export type MarketMoodFreshnessStatus = "fresh" | "stale" | "unavailable";

export type MarketMoodRating =
  | "extreme_fear"
  | "fear"
  | "neutral"
  | "greed"
  | "extreme_greed"
  | "unknown";

export type MarketMoodComparisonWindow = "1w" | "1m" | "1y";

export type MarketMoodRefreshStatus = "refreshed" | "unchanged" | "failed";

export type MarketMoodAxisValueFormat =
  | "number"
  | "percent"
  | "ratio"
  | "index"
  | "currency"
  | "spread"
  | "unknown";

export type MarketMoodValueMeaning = "fear" | "greed" | "neutral_or_contextual" | "unknown";

/* ── Nested reads ──────────────────────────────────────────────────────── */

export interface MarketMoodTrendPointRead {
  date: string;
  score: number | null;
  score_label: string | null;
  rating: MarketMoodRating;
  rating_label: string;
}

export interface MarketMoodComparisonRead {
  window: MarketMoodComparisonWindow;
  prior_score: number | null;
  prior_score_label: string | null;
  change_label: string | null;
  is_available: boolean;
}

export interface MarketMoodComponentRead {
  component_key: string;
  display_name: string;
  score: number | null;
  score_label: string | null;
  rating: MarketMoodRating;
  rating_label: string;
}

/* ── Detail reads (per-indicator history) ──────────────────────────────── */

export interface MarketMoodIndicatorHistoryPointRead {
  date: string;
  /** Raw indicator value (provider/native scale); presentational only. */
  value: number | null;
  /** Backend-formatted display label for the raw value; render verbatim. */
  value_label: string | null;
  /** Normalized 0–100 sub-score for this point (secondary context). */
  score: number | null;
  score_label: string | null;
  rating: MarketMoodRating;
  rating_label: string;
}

export interface MarketMoodIndicatorRead {
  component_key: string;
  display_name: string;
  subtitle: string;
  description: string;
  current_score: number | null;
  current_score_label: string | null;
  current_rating: MarketMoodRating;
  current_rating_label: string;
  current_value: number | null;
  current_value_label: string | null;
  unit_label: string | null;
  axis_label: string | null;
  axis_value_format: MarketMoodAxisValueFormat;
  higher_value_meaning: MarketMoodValueMeaning;
  lower_value_meaning: MarketMoodValueMeaning;
  history: MarketMoodIndicatorHistoryPointRead[];
}

/* ── Top-level read ────────────────────────────────────────────────────── */

export interface MarketMoodRead {
  data_mode: MarketMoodDataMode;
  source_label: string;
  source_detail_label: string;
  source_rights_notice: string;
  /** ISO 8601 datetime string from backend. */
  generated_at: string;
  /** ISO 8601 datetime string from backend, or null when unknown. */
  updated_at_utc: string | null;
  updated_at_label: string | null;
  freshness_status: MarketMoodFreshnessStatus;
  freshness_label: string;
  is_trading_signal: boolean;
  is_actionability_input: boolean;
  is_risk_rule_input: boolean;
  score: number | null;
  score_label: string | null;
  score_min: number;
  score_max: number;
  rating: MarketMoodRating;
  rating_label: string;
  trend_series: MarketMoodTrendPointRead[];
  comparisons: MarketMoodComparisonRead[];
  components: MarketMoodComponentRead[];
  caveat_codes: string[];
  limitations: string[];
  status_message: string | null;
}

/* ── Detail read (overall + 7 indicators with raw-scale history) ───────── */

export interface MarketMoodDetailRead {
  data_mode: MarketMoodDataMode;
  source_label: string;
  source_detail_label: string;
  source_rights_notice: string;
  generated_at: string;
  updated_at_utc: string | null;
  updated_at_label: string | null;
  freshness_status: MarketMoodFreshnessStatus;
  freshness_label: string;
  is_trading_signal: boolean;
  is_actionability_input: boolean;
  is_risk_rule_input: boolean;
  score: number | null;
  score_label: string | null;
  score_min: number;
  score_max: number;
  rating: MarketMoodRating;
  rating_label: string;
  trend_series: MarketMoodTrendPointRead[];
  comparisons: MarketMoodComparisonRead[];
  indicators: MarketMoodIndicatorRead[];
  caveat_codes: string[];
  limitations: string[];
  status_message: string | null;
}

/* ── Refresh status (manual/backend-only; not consumed by the dashboard card) ─ */

export interface MarketMoodRefreshStatusRead {
  status: MarketMoodRefreshStatus;
  data_mode: MarketMoodDataMode;
  source_label: string;
  generated_at: string | null;
  updated_at_utc: string | null;
  source_changed: boolean | null;
  last_checked_at_utc: string | null;
  last_checked_at_label: string | null;
  record_count: number;
  message: string;
}
